### ISSUE TITLE
Feature request: expose isPlaying state through return value of audix()

### ISSUE THREAD

**Issue by marianneke** [keli-wen/streamlit-advanced-audio#1]
For my use case, it would be very helpful if I have access to the state of isPlaying from the audix widget, in order to control some other behavior that depends on whether audio is playing. Is it possible to expose this state through the return value of audix()?

**Comment by keli-wen** [comment]
I believe I can easily include the current state of the widget in the
return value.

I'll make the necessary modifications, and the update should be ready
within the next 2-3 days.

Thank you for your feedback! ✨

Marianne Hoogeveen ***@***.***> 于2024年12月31日周二 05:11写道：

**Comment by keli-wen** [comment]
Almost forgot to say: Happy New Year! 🎉

文柯力 ***@***.***> 于2025年1月2日周四 16:32写道：

**Comment by marianneke** [comment]
> Almost forgot to say: Happy New Year! 🎉 文柯力 ***@***.***> 于2025年1月2日周四 16:32写道：

Happy New Year to you too! 
Thanks for the prompt response!!! This would be huge in terms of allowing me to clean up my code!

Also, love your avatar :fish_cake:

**Comment by keli-wen** [comment]
I've resolved the issue in this commit: https://github.com/keli-wen/streamlit-advanced-audio/commit/465f345a75dac2db36b88bce464ee1aa97c889d7. I've deployed the updated repository to Streamlit Cloud. You can try out the new feature on the website linked from my repository's homepage or read the updated [README.md](https://github.com/keli-wen/streamlit-advanced-audio?tab=readme-ov-file#basic-usage). ✨

**Comment by marianneke** [comment]
> I've resolved the issue in this commit: [465f345](https://github.com/keli-wen/streamlit-advanced-audio/commit/465f345a75dac2db36b88bce464ee1aa97c889d7). I've deployed the updated repository to Streamlit Cloud. You can try out the new feature on the website linked from my repository's homepage or read the updated [README.md](https://github.com/keli-wen/streamlit-advanced-audio?tab=readme-ov-file#basic-usage). ✨

Amazing!
Do I need to use WaveSurferOptions in order to access the isPlaying field, or can I simply do `result = audix(audio)` and access this information?

**Comment by keli-wen** [comment]
Yes, and you can refer to this demo: https://advanced-audio-example.streamlit.app/ or the `example.py` python file. If you found `audix` really helpful, maybe you could even give it a star Thank you!

**Comment by jersonjunior** [comment]
Firstly, congratulations on the project! 
Would it be possible for the audix() function to trigger automatic updates while the audio is playing?

I need to synchronize the audio time with the transcribed text and as the audio plays the text appears.

**Comment by keli-wen** [comment]
> Firstly, congratulations on the project! Would it be possible for the audix() function to trigger automatic updates while the audio is playing?
> 
> I need to synchronize the audio time with the transcribed text and as the audio plays the text appears.

Thanks for your attention!

Initially, I designed it this way, but I might have encountered a design issue. If I update the return value while playing (similar to [this part of the code](https://github.com/keli-wen/streamlit-advanced-audio/blob/master/streamlit_advanced_audio/frontend/src/Audix.tsx#L282-L286) but adding a return function after it), it causes severe performance issues for slightly longer audio.

My understanding is that dynamically updating the return value triggers Streamlit to rerun the page repeatedly, leading to performance problems.

Let me know if you have any suggestions or ideas!

**Comment by jersonjunior** [comment]
> > Firstly, congratulations on the project! Would it be possible for the audix() function to trigger automatic updates while the audio is playing?
> > I need to synchronize the audio time with the transcribed text and as the audio plays the text appears.
> 
> Thanks for your attention!
> 
> Initially, I designed it this way, but I might have encountered a design issue. If I update the return value while playing (similar to [this part of the code](https://github.com/keli-wen/streamlit-advanced-audio/blob/master/streamlit_advanced_audio/frontend/src/Audix.tsx#L282-L286) but adding a return function after it), it causes severe performance issues for slightly longer audio.
> 
> My understanding is that dynamically updating the return value triggers Streamlit to rerun the page repeatedly, leading to performance problems.
> 
> Let me know if you have any suggestions or ideas!

The interesting thing is that when I press pause I receive the current_time status and can view the transcribed audio text, however I have to keep clicking the play and pause button. Anyway, I'll think of another way to solve this. 

Thank you for your attention.

**Comment by keli-wen** [comment]
> > > Firstly, congratulations on the project! Would it be possible for the audix() function to trigger automatic updates while the audio is playing?
> > > I need to synchronize the audio time with the transcribed text and as the audio plays the text appears.
> > 
> > 
> > Thanks for your attention!
> > Initially, I designed it this way, but I might have encountered a design issue. If I update the return value while playing (similar to [this part of the code](https://github.com/keli-wen/streamlit-advanced-audio/blob/master/streamlit_advanced_audio/frontend/src/Audix.tsx#L282-L286) but adding a return function after it), it causes severe performance issues for slightly longer audio.
> > My understanding is that dynamically updating the return value triggers Streamlit to rerun the page repeatedly, leading to performance problems.
> > Let me know if you have any suggestions or ideas!
> 
> The interesting thing is that when I press pause I receive the current_time status and can view the transcribed audio text, however I have to keep clicking the play and pause button. Anyway, I'll think of another way to solve this.
> 
> Thank you for your attention.

This is my design. When you drag the timeline (seeking) and start/pause playback, I will have `audix` perform a status update (returning the current time). I'm considering supporting a **minimal updating interval** to control the frequency of status updates (I tried this before, but failed, and I will continue to try).

**Comment by jersonjunior** [comment]
> > > > Primeiramente, parabéns pelo projeto! Seria possível a função audix() disparar atualizações automáticas enquanto o áudio está tocando? 
> > > > Preciso sincronizar o tempo do áudio com o texto transcrito e conforme o áudio toca o texto aparece.
> > > 
> > > 
> > > Obrigado pela atenção! 
> > > Inicialmente, eu projetei dessa forma, mas posso ter encontrado um problema de design. Se eu atualizar o valor de retorno durante a reprodução (semelhante a [esta parte do código](https://github.com/keli-wen/streamlit-advanced-audio/blob/master/streamlit_advanced_audio/frontend/src/Audix.tsx#L282-L286) , mas adicionando uma função de retorno depois dela), isso causa sérios problemas de desempenho para áudio um pouco mais longo. 
> > > Meu entendimento é que atualizar dinamicamente o valor de retorno aciona o Streamlit para executar a página repetidamente, levando a problemas de desempenho. 
> > > Me avise se você tiver alguma sugestão ou ideia!
> > 
> > 
> > O interessante é que quando eu aperto pause eu recebo o status current_time e consigo visualizar o texto do áudio transcrito, porém tenho que ficar clicando no botão play e pause. De qualquer forma, vou pensar em outra maneira de resolver isso.
> > Obrigado pela sua atenção.
> 
> Este é meu design. Quando você arrasta a linha do tempo (buscando) e inicia/pausa a reprodução, eu terei que `audix`executar uma atualização de status (retornando o tempo atual). Estou considerando oferecer suporte a um **intervalo mínimo de atualização** para controlar a frequência das atualizações de status (eu tentei isso antes, mas falhei, e continuarei tentando).

Look this example: https://wavesurfer.xyz/examples/?react.js
